Compilation warnings in lib/openssl.c

Compiling the most recent version () on windows, I get the following warnings:

~~~
11>------ Build started: Project: picotlsvs, Configuration: Release x64 ------
...
9>Generating code
10>c:\users\huitema\github\picotls\lib\openssl.c(1099): warning C4311: 'type cast': pointer truncation from 'HANDLE' to 'int'
10>picotls.c
10>c:\users\huitema\github\picotls\t\picotls.c(1227): warning C4312: 'type cast': conversion from 'unsigned int' to 'void (__cdecl *)(ptls_async_job_t *)' of greater size
11>Generating code
~~~

The code implicitly assume that a "handle" (often a pointer) is the same bit size as an "int". For example in openssl.c we see:
~~~
int async_sign_ctx_get_fd(ptls_async_job_t *_self)
{
    struct async_sign_ctx *self = (void *)_self;
    OSSL_ASYNC_FD fds[1];
    size_t numfds;

    ASYNC_WAIT_CTX_get_all_fds(self->waitctx, NULL, &numfds);
    assert(numfds == 1);
    ASYNC_WAIT_CTX_get_all_fds(self->waitctx, fds, &numfds);
    return (int)fds[0];
}
~~~
`fds[0]` is of type `OSSL_ASYNC_FD`, which on Windows is defined as `HANDLE`, which on 64 bits build of windows is a 64 bit number. The `(int)` cast will only retain the last 32 bits.

--- a/include/picotls.h
+++ b/include/picotls.h
@@ -779,7 +779,7 @@ typedef struct st_ptls_async_job_t {
     /**
      * optional callback returning a file descriptor that becomes readable when the job is complete
      */
-    int (*get_fd)(struct st_ptls_async_job_t *self);
+    intptr_t (*get_fd)(struct st_ptls_async_job_t *self);
     /**
      * optional callback for setting a completion callback
      */

--- a/lib/openssl.c
+++ b/lib/openssl.c
@@ -1088,7 +1088,7 @@ static void async_sign_ctx_free(ptls_async_job_t *_self)
     free(self);
 }
 
-int async_sign_ctx_get_fd(ptls_async_job_t *_self)
+intptr_t async_sign_ctx_get_fd(ptls_async_job_t *_self)
 {
     struct async_sign_ctx *self = (void *)_self;
     OSSL_ASYNC_FD fds[1];
@@ -1097,7 +1097,7 @@ int async_sign_ctx_get_fd(ptls_async_job_t *_self)
     ASYNC_WAIT_CTX_get_all_fds(self->waitctx, NULL, &numfds);
     assert(numfds == 1);
     ASYNC_WAIT_CTX_get_all_fds(self->waitctx, fds, &numfds);
-    return (int)fds[0];
+    return (intptr_t)fds[0];
 }
 
 static ptls_async_job_t *async_sign_ctx_new(const ptls_openssl_signature_scheme_t *scheme, EVP_MD_CTX *ctx, size_t siglen)

--- a/t/openssl.c
+++ b/t/openssl.c
@@ -482,7 +482,7 @@ static void many_handshakes(void)
             case PTLS_ERROR_ASYNC_OPERATION: {
                 ptls_async_job_t *job = ptls_get_async_job(qat.conns[offending].tls);
                 assert(job->get_fd != NULL);
-                qat.conns[offending].wait_fd = job->get_fd(job);
+                qat.conns[offending].wait_fd = (int)job->get_fd(job);
                 assert(qat.conns[offending].wait_fd != -1);
             } break;
             default:

--- a/t/picotls.c
+++ b/t/picotls.c
@@ -1220,7 +1220,7 @@ static int async_sign_certificate(ptls_sign_certificate_t *self, ptls_t *tls, pt
             int ret = sign_certificate(self, tls, NULL, selected_algorithm, &fakebuf, input, algorithms, num_algorithms);
             assert(ret == 0);
             ptls_buffer_dispose(&fakebuf);
-            async_ctx.super.destroy_ = (void (*)(ptls_async_job_t *))0xdeadbeef;
+            async_ctx.super.destroy_ = NULL;
             async_ctx.selected_algorithm = *selected_algorithm;
             *async = &async_ctx.super;
             --server_sc_callcnt;

